### PR TITLE
Add a context for making things

### PR DESCRIPTION
**Pull request by WaffleLapkin**
r? @oli-obk 
cc @nnethercote (I've touched these files recently)

TL;DR: replace `tcx.mk_something()` with `tcx.mk().something()` in order to improve documentation (less stuff in the `TyCtxt` impls)

Future work:
- [ ] Remove `mk_*` methods from `TyCtxtAt` maybe
- [ ] Better document `MkCtxt`'s methods
- [ ] Move methods that are missing `mk_` or have `_mk_` in the middle

-----

I'm not 100% sure this is a good idea, but generally de-bloating `TyCtxt` seems good.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context/mk.rs [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+}
+
+impl<'tcx> TyCtxt<'tcx> {
+    /// Returns the "make context" with some extension methods.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust,no_run
+    /// # rustc_middle::ty::TyCtxt;
+    /// let tcx: TyCtxt = todo!();
```
Comment (Nilstrieb) [inline compiler/rustc_middle/src/ty/context/mk.rs:45]:
```suggestion
    /// # let tcx: TyCtxt = todo!();
```

**Comment by compiler-errors** [comment]
Huh -- not sure if I like this better than just normalizing all the method names (make sure they start with `mk_`, like `mk_ty_projection`) and moving them to a separate module (`ty/mk.rs` here).Basically I'm not a huge fan of having to add `.mk()` on every invocation.

**Comment by oli-obk** [comment]
I do like it 😃

But it does seem affect enough code to warrant an MCP (and then compiler-errors and I can fight it out in a zulip thread 😃).

The MCP should probably be more about the general idea, as it seems applicable to other things, too

**Comment by nnethercote** [comment]
`mk_foo` is a verb phrase with an obvious meaning, which is good for function names. `mk().foo()` is... not that. `mk` should really be called `mcx` to follow the existing convention for things like `tcx` and `icx`, but `mcx().foo()` doesn't work. I feel like this cure is worse than the disease.

**Comment by oli-obk** [comment]
We could use const generics of `Symbol` type and do `tcx.mk<thing>(args)`. Tho that would mean multiple args would need to get tupled. Tho at that point we can go all the way and infer the (tupled if multiple) args from the return type and just have `tcx.mk(args)`

**Comment by WaffleLapkin** [comment]
@oli-obk @compiler-errors I've opened an MCP, now fight https://github.com/rust-lang/compiler-team/issues/603

**Comment by WaffleLapkin** [comment]
Closed as the MCP was not accepted.

**Pull request closed without merging**